### ISSUE TITLE
Trigger from outside flow?

### ISSUE THREAD

**Issue by rivertam** [DCtheTall/twilly#3]
So I want to create a flow and then have a website form trigger that flow for a phone number entered in the form. Should I just use twilio's normal API? This seems like an oddly missing part of twilly.

**Comment by DCtheTall** [comment]
This is a great idea! I admit this is definitely missing key functionality.

I am going to look into adding this (for today I will add it to the "Feature Pipeline" section of the README). Right now I imagine that any entry point should take a FlowSchema as a parameter for robustness.

PRs are welcome if you are interested in working on it, just let me know if you plan on working on it!

**Comment by DCtheTall** [comment]
Leaving some notes here.

**Q:** The reason all Flows start from a request from Twilio is because Twilly sets an [SMS cookie](https://support.twilio.com/hc/en-us/articles/223136287-How-do-Twilio-cookies-work-) which stores the SMS cookie, which is what is used to store the interaction's state in the graph. How do we get the cookie for the remote trigger?

**A:**  We can make some dummy GET request to Twilio's API and read the cookie from their response.

So far there are a couple options to how to introduce this. Since Twilly returns an Express router, the `twilly()` function's interface can be modified so that you can choose API endpoints which trigger Actions. This leans toward the use case where the server with the Twilly code can be used as an SMS application micro-service.

If you want to trigger a Flow outside of a webhook, and remotely during some asynchronous Node.js procedure, then there are 2 options:

1. Make breaking changes to Twilly's API and do a major release. This would allow me to make a more intuitive interface where the `twilly()` function returns an object where one key is a Router and the other key is a function which takes a Flow's name (string) as a parameter and will trigger the Flow.

2. Have the `twilly()` function accept an additional parameter: a callback which you can use as a closure to set one of the parameters passed to the callback to a higher-scoped variable.

Since the 2nd pattern is more complex, I am leaning towards the first option, but do not want to make breaking changes lightly.

**Comment by rivertam** [comment]
So I know very little about twilly internals, but I think it would be possible to pick any number available to the service, right? I'll say at least for my usecase, it's actually *preferable* if the `from` does not have an associated cookie, because I might *prefer* to use a number the user has never seen before if I'm the one initiating contact. That way, at least theoretically, the user can have multiple conversations with the service at the same time. I just want to point that out there -- the SMS cookie looks like a nice signal, but I'm not sure you're treating it right ("handle the case" rather than "prefer the case", for example).

I won't speak for other users, but just saying if I trigger I definitely don't need you to do anything terribly special in the case that the cookie has expired; that's probably the way I wanted it anyways. That said, I wouldn't necessarily expect twilly to pick a number the user's never seen before. As long as it's documented, any number is fine.

**Comment by DCtheTall** [comment]
@rivertam I have started implementing this feature and I found that my comment was actually not correct. The SMS cookie is stored in Twilio's server and is sent with requests, but not responses (AFAIK, I could be wrong). So we cannot get the SMS cookie unless a Twilio server makes a request to the Express app. So it will start with a fresh state each time.

My goal is to have a `triggerFlow()` function  instead of creating an instance of a Trigger action since Action objects are actually mostly read-only. The function will take a Flow name and a from number.

Unfortunately I cannot think of a way to make this feature without a minor regression, you won't be able to have multiple Twilly routers in your app without providing extra parameters to the `twilly()` function call. Since I expect this use case to be rare, I think it is worth it though.

I'll update the docs on how to use multiple Twilly routers with different FlowSchema's once this change is merged into master and released.

**Comment by DCtheTall** [comment]
Just an update: I am working on this, the implementation is done. I just need to:

1. Test it with my own Twilio account.
2. Test that the changes I made did not cause a regression.
3. Add unit tests for the new functionality.

**Comment by DCtheTall** [comment]
@rivertam check out [this PR](https://github.com/DCtheTall/twilly/pull/7) if you want to get a sense of what it will look like. Unfortunately there is no way to use Question actions in the Flow, and you cannot trigger other Flows, but it provides basic messaging functionality. Any state you want to persist across `triggerFlow` and your larger FlowSchema should be accessible through the `getUserContext` hook.

**Comment by rivertam** [comment]
Sorry I've been AWOL. I actually realized it might be illegal in some places to message people from a service without prompting. It's obviously useful in some cases, but for my use-case it seems too risky, so I'm changing my application to be web-first rather than a messaging service. Unfortunately, that means I can't really opine on this topic right now.

**Comment by DCtheTall** [comment]
@rivertam no worries, I have been busy with work/travel recently as well. I have a [PR](https://github.com/DCtheTall/twilly/pull/7) with the feature that is still missing tests. When I get around to those I will merge it to master, since I think this use case is generally useful given my own experience developing w Twilio.

**Comment by DCtheTall** [comment]
@rivertam FWIW if you wanted to implement this without having to wait on me merging #7 (which I will get to someday) you could also use Twilio's Node.js API to send the text, then use your Twilly graph to just handle the user's reply.
